# **Notebook 5 – Probability**

## **1. Introduction to Probability**

**Understand the Concept**

* Probability tells us how likely something is to happen.
* Its value is between 0 & 1.
* 0 - event cannot happen.
* 1 - event will definitely happen.

**Formula :** P(E) = Number of Favorable Outcomes / Total Number of Outcomes

**Demonstrate the Concept**

**Example:**

* A dice has 6 possible outcomes: 1, 2, 3, 4, 5, 6.
* We want to find the probability of getting 6.
* Favorable outcomes = 1
* Total outcomes = 6
* Probability = 1 / 6 = 0.1667

In [1]:
# Implement the Concept:
# Dice - 6 possible outcomes
total_outcomes = 6

# 1 favorable outcome 
favorable_outcomes = 1

# Probability 
probability = favorable_outcomes / total_outcomes

print("Total Outcomes:", total_outcomes)
print("Favorable Outcomes:", favorable_outcomes)
print("Probability of getting 6:", probability)

Total Outcomes: 6
Favorable Outcomes: 1
Probability of getting 6: 0.16666666666666666


So, the probability of getting 6 is 16.67%.

## **2. Events**

**Understand the Concept**

* An event is an outcome or group of outcomes.
* It can be something that happens or does not happen.
* It can have one or more possible outcomes.


**Demonstrate the Concept**

**Example:**

* We roll a dice once.
* The event is "getting an even number."
* Even numbers : 2, 4, and 6.
* Favorable outcomes = 3
* Total outcomes = 6
* Probability = 3 / 6 = 0.5

In [2]:
# Implement the Concept:

total_outcomes = 6
favorable_outcomes = 3

# Find probability
probability = favorable_outcomes / total_outcomes

print("Favorable Outcomes:", favorable_outcomes)
print("Total Outcomes:", total_outcomes)
print("Probability of getting an even number:", probability)

Favorable Outcomes: 3
Total Outcomes: 6
Probability of getting an even number: 0.5


So, the probability of getting an even number is 50%.

## **3. Sample Space**

**Understand the Concept**

* Sample space means all possible outcomes of an experiment.



**Demonstrate the Concept**

**Example:**

* We roll a dice once.
* The possible outcomes : 1, 2, 3, 4, 5, 6.
* So, the sample space is {1, 2, 3, 4, 5, 6}.


In [3]:
# Implement the Concept:

# Possible outcomes when rolling a dice
sample_space = {1, 2, 3, 4, 5, 6}

print("Sample Space:", sample_space)
print("No of Possible Outcomes:", len(sample_space))

Sample Space: {1, 2, 3, 4, 5, 6}
No of Possible Outcomes: 6


## **4. Independent Events**

**Understand the Concept**

* Independent events are events where one event does not affect the other event.
* The result of one event does not change the probability of the second event.

**Formula :** P(A and B) = P(A) × P(B)

**Demonstrate the Concept**

**Example:**

* We toss a coin and roll a dice.
* Getting Head on the coin does not affect the result of the dice.
* Probability of Head = 1 / 2
* Probability of getting 6 = 1 / 6
* Probability of getting Head and 6 = 1 / 2 × 1 / 6 = 1 / 12

In [4]:
# Implement the Concept:
# 2 events
probability_head = 1 / 2
probability_six = 1 / 6

# Probability
probability_both = probability_head * probability_six

print("Probability of Head:", probability_head)
print("Probability of 6:", probability_six)
print("Probability of Head and 6:", probability_both)

Probability of Head: 0.5
Probability of 6: 0.16666666666666666
Probability of Head and 6: 0.08333333333333333


The probability of getting Head and 6 is 8.33%.

## **5. Dependent Events**

**Understand the Concept**

* Dependent events are events where one event affects the other event.
* The result of the first event changes the probability of the second event.

**Formula :** P(A and B) = P(A) × P(B | A)

**Demonstrate the Concept**

**Example:**

* A bag has 4 red apples and 6 green apples.
* We pick one red apple and do not put it back.
* Now, 3 red apples are left out of 9 apples.
* So, the probability of picking a red apple the second time changes.

In [5]:
# Implement the Concept:
# Apples in bag
red_apples = 4
green_apples = 6
total_apples = 10

# 1st event - pick one red apple
first_event = red_apples / total_apples  # P(A)

# Now 1 red apple is removed in bag
current_red_apples = red_apples - 1 # 3 r apple
current_total_apples = total_apples - 1 # 9 tot apple

# 2nd event - pick another 1 red apple
second_event = current_red_apples / current_total_apples # P(B|A)

# Dependent event 
probability_1_and_2 = first_event * second_event # P(A and B) 

print("1st event probability : ",first_event )
print("2nd event probability of both event : ",second_event )
print("probability of both event : ",probability_1_and_2)

1st event probability :  0.4
2nd event probability of both event :  0.3333333333333333
probability of both event :  0.13333333333333333


The probability of picking two red apples is 13.33%.

## **6. Conditional Probability**

**Understand the Concept**

* Conditional probability tells us the probability of an event when we already know that another event has happened.
* It helps us find the chance of one event under a given condition.

**Formula :** P(A | B) = P(A and B) / P(B)



**Demonstrate the Concept**

**Example:**

* A class has 20 students.
* 12 students play cricket.
* 8 students play both cricket and football.
* We want to find the probability that a student plays football given that the student plays cricket.
* P(Football | Cricket) = 8 / 12 = 0.67


In [6]:
# Implement the Concept:
# Students play cricket
cricket_students = 12

# Play both cricket and football
both_sports = 8

# Conditional probability
probability_football_given_cricket = both_sports / cricket_students
print("P(Football | Cricket):", probability_football_given_cricket)

P(Football | Cricket): 0.6666666666666666


So, the probability of playing football given that the student plays cricket is 66.77%.

## **7. Bayes' Theorem**

**Understand the Concept**

* Bayes' Theorem helps us find the probability of an event using new information.
* It updates an earlier probability when we get new evidence.
* It is useful when we know the result of one event and want to find the chance of another event.

**Formula :** P(A | B) = [P(B | A) × P(A)] / P(B)



**Demonstrate the Concept**

**Example:**

* Suppose 30% of emails are spam.
* 80% of spam emails contain the word "Offer".
* 20% of normal emails also contain the word "Offer".
* We want to find the probability that an email is spam when it contains the word "Offer".

In [7]:
# Implement the Concept:
# Bayes' Theorem:  P(Spam | Offer) = (P(Offer | Spam) * P(Spam)) / P(Offer)

p_spam = 0.30
p_offer_spam = 0.80
p_offer_normal = 0.20
p_normal = 1 - p_spam

# Probability of getting "Offer" in both spam & normal
p_offer = (p_offer_spam * p_spam + p_offer_normal * p_normal)

# Bayes' Theorem
p_spam_given_offer = (p_offer_spam * p_spam) / p_offer

print("P(Offer):", p_offer)
print("P(Spam | Offer):", p_spam_given_offer)

P(Offer): 0.38
P(Spam | Offer): 0.631578947368421


* The probability of getting the word "Offer" is 38%.
* If an email contains the word "Offer", the chance that it is spam is about 63.16%.

## **8. Probability Rules**

**Understand the Concept**

* Probability follows some basic rules.
* The probability of an event is always between 0 and 1.
* The probability of the complete sample space is 1.
* The probability of an event not happening is 1 minus its probability.

**Basic Rules**

**Rule 1:** 0 ≤ P(E) ≤ 1

**Rule 2:** P(Sample Space) = 1

**Rule 3:** P(Not A) = 1 - P(A)



**Demonstrate the Concept**

**Example:**

* We roll a dice once.
* Even numbers are 2, 4, and 6.
* Probability of getting an even number = 3/6 = 0.5.
* Probability of not getting an even number = 1 - 0.5 = 0.5.



In [8]:
# Implement the Concept:
total_outcomes = 6
even_outcomes = 3

# Probability of getting an even number
p_even = even_outcomes / total_outcomes

# Probability of not getting an even number
p_not_even = 1 - p_even

print("P(Even):", p_even)
print("P(Not Even):", p_not_even)
print("P(Sample Space):", 1)

P(Even): 0.5
P(Not Even): 0.5
P(Sample Space): 1


## **9. Random Variables**

**Understand the Concept**

* A random variable is a variable that represents the result of a random event.
* It can take different values based on the outcome.
* It can be a number or a category represented by numbers.

**Demonstrate the Concept**

**Example:**

* We roll a dice once.
* Let X represent the number we get.
* X possible values : 1, 2, 3, 4, 5, 6.


In [9]:
# Implement the Concept:
import numpy as np

# Random values of X
X = np.random.randint(1, 7) 

print("Dice Roll:", X)

Dice Roll: 4


X value are differ when roll a dices every time 

## **10. Expected Value**

**Understand the Concept**

* Expected value is the average result we can expect from a random event over many times.

**Formula :** Expected Value = Σ (Value × Probability)



**Demonstrate the Concept**

**Example:**

* A game gives ₹100 for a win and ₹0 for a loss.
* Probability of winning = 0.5
* Probability of losing = 0.5

Expected Value = (100 × 0.5) + (0 × 0.5)
Expected Value = ₹50


In [10]:
# Implement the Concept:
# Values
win_amount = 100
loss_amount = 0

# Probability
win_probability = 1/2
loss_probability = 1/2

# Formula : Expected Value = Σ (Value × Probability)
expected_value = (win_amount * win_probability) + (loss_amount * loss_probability)

print("Expected Value:", expected_value)

Expected Value: 50.0


**Explaination :**

* The expected value is ₹50.
* This does not mean we will always win ₹50.
* It means that over many similar games, the average result would be around ₹50 per game.

## Interview Questions & Answers

**Q1. What is Probability?**

**Answer:**

Probability tells us how likely something is to happen. Its value is between 0 and 1.

**Q2. What is an Event?**

**Answer:**

An event is an outcome or group of outcomes that we are interested in.

**Q3. What is Sample Space?**

**Answer:**

Sample space is the list of all possible outcomes of an experiment.

**Q4. What are Independent Events?**

**Answer:**

Independent events are events where one event does not affect the other event.

**Q5. What are Dependent Events?**

**Answer:**

Dependent events are events where one event affects the other event.

**Q6. What is Conditional Probability?**

**Answer:**

Conditional probability tells us the chance of an event when we already know another event has happened.

**Q7. What is Bayes' Theorem?**

**Answer:**

Bayes' Theorem helps us update probability using new information or evidence.

**Q8. What is Expected Value?**

**Answer:**

Expected value is the average result we can expect over many repeated trials.